url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/util/SourceGenerator.java

```diff
@@ -0,0 +1,216 @@
+package fri.patterns.interpreter.parsergenerator.util;
+
+import java.io.*;
+import java.util.*;
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.parsertables.*;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+import fri.patterns.interpreter.parsergenerator.syntax.builder.SyntaxBuilder;
+
+/**
+	Java source generator. Following code generations are supported:
+	<ul>
+		<li>SemanticSkeleton, code base for a Semantic-implementation of big syntaxes.</li>
+		<li>Turn ParserTables (built from a Syntax) into compileable Java-code (for faster loading).</li>
+		<li>Turn a Syntax object into compileable Java code (for faster loading).</li>
+	</ul>
+	<pre>
+	SYNTAX: java fri.patterns.interpreter.parsergenerator.util.SourceGenerator [semantic|LALR|SLR|LR] file.syntax [file.syntax ...]
+			LALR|SLR|LR: Generates ParserTable implementation(s) of passed grammar file(s).
+			       else: Generates Syntax implementation(s) of passed grammar file(s).
+	CAUTION: Files MUST have relative pathes!
+	</pre>
+
+	@author (c) 2002, Fritz Ritzberger
+*/
+
+public abstract class SourceGenerator
+{
+	/**
+		Generates a semantic skeleton implementation for a given syntax.
+		@param input syntax the semantic is meant for
+		@param className basename of class to generate, semantic will be named className+"Semantic.java"
+		@param pkgName package-name of class to generate
+	*/
+	public static void generateSemanticSkeleton(Syntax syntax, String className, String pkgName)
+		throws Exception
+	{
+		String fileName = className+"Semantic.java";
+
+		String dirName = pkgName != null && pkgName.length() > 0
+				? pkgName.replace('.', File.separatorChar)
+				: System.getProperty("user.dir");
+
+		File out = new File(dirName, fileName);
+				
+		if (out.exists())	{
+			throw new IllegalStateException("Will not overwrite "+out.getAbsolutePath()+". Please check the file for implementation and remove it!");
+		}
+		else	{
+			new File(dirName).mkdirs();	// ensure directory exists
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Confirm that `mkdirs` was successful. Handle any failures that happened. If you fail silently, then you might encounter errors that are harder to debug. See the following diffs on GitHub that make this code change: [diff1](https://github.com/qunarcorp/qmq/commit/12f1d9d6a57a02bc2415cb2af9f4388a77b12c02#diff-6b67be6e17f24fdeddb7c725ac8ec5600d6e63be0bdca0963a83d530e6d55e35L36), [diff2](https://github.com/apache/jmeter/commit/1ec61206779b3f84df18516682b4791bf015178f#diff-cd74304763c097fbc085ac2bc952b46ab3843ac825cacacad89e8064296e955aL330)

Learn more: [File.mkdirs in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdirs--)